### PR TITLE
Card

### PR DESCRIPTION
**Pull request by DevIgork**

### CODE REVIEW

File: src/main/java/com/example/bookstore/dto/shopingcart/AddToCartRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package com.example.bookstore.dto.shopingcart;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+import lombok.Data;
+
+@Data
+public class AddToCartRequestDto {
+    @NotNull
```
Comment (nshtykh) [inline src/main/java/com/example/bookstore/dto/shopingcart/AddToCartRequestDto.java:9]: You can use @Positive here too

**Review by nshtykh — APPROVED** [review]
Great job!

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    @Operation(summary = "update cartItem", description = "update cart item by Id")
+    @PutMapping("/cart-items/{cartItemId}")
+    public CartItemDto updateCart(
+            @PathVariable Long cartItemId,
+            @Valid @RequestBody UpdateCartItemRequestDto requestDto
+    ) {
+        return shoppingCartService.update(cartItemId, requestDto);
+    }
+
+    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER')")
```

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    @Operation(summary = "Get cart", description = "Get a user cart")
+    @GetMapping
+    public ShoppingCartDto getCart(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.get(user.getId());
+    }
+
+    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER')")
```

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    @PostMapping
+    public ShoppingCartDto addToCart(
+            @RequestBody @Valid AddToCartRequestDto requestDto,
+            Authentication authentication
+    ) {
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.addToCart(requestDto, user.getId());
+    }
+
+    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER')")
```

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+@Tag(name = "Shopping cart management", description = "Endpoints for managing shopping cart")
+@RestController
+@RequestMapping("/api/cart")
+@RequiredArgsConstructor
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @PreAuthorize("hasAnyAuthority('ROLE_USER', 'ROLE_ADMIN')")
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java]:
```suggestion
    @PreAuthorize("hasAnyAuthority('ROLE_USER')")
```

File: src/main/java/com/example/bookstore/controller/ShoppingCartController.java
Comment (d1sam) [inline src/main/java/com/example/bookstore/controller/ShoppingCartController.java:1]: Available endpoints for users with role ADMIN

- POST: /api/books/
- PUT: /api/books/{id}
- DELETE: /api/books/{id}
- POST: /api/categories
- PUT: /api/categories/{id}
- DELETE: /api/categories/{id}

File: src/main/java/com/example/bookstore/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/bookstore/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/bookstore/model/ShoppingCart.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+@SQLDelete(sql = "UPDATE shopping_cart SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @OneToOne
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/model/ShoppingCart.java]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/com/example/bookstore/service/impl/ShopingCartServiceIml.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                });
+        cartItem.setQuantity(requestDto.getQuantity());
+        cartItem.setBook(book);
+        cartItem.setShoppingCart(shoppingCartFromDB);
+        cartItemRepository.save(cartItem);
+        shoppingCartFromDB.getCartItem().add(cartItem);
+        return shoppingCartMapper.toDto(shoppingCartFromDB);
+    }
+
+    @Transactional
```
Comment (d1sam) [inline src/main/java/com/example/bookstore/service/impl/ShopingCartServiceIml.java]:
```suggestion
    @Transactional(readOnly = true)
```

**Review by d1sam — CHANGES_REQUESTED** [review]
Good job, but there are some changes should be done!

**Review by yevheniisynytsia — APPROVED** [review]
:fire:

**Pull request merged**